planner/core/rule_predicate_push_down.go

```diff
@@ -726,3 +742,123 @@ func appendAddSelectionTraceStep(p LogicalPlan, child LogicalPlan, sel *LogicalS
 	}
 	opt.appendStepToCurrent(sel.ID(), sel.TP(), reason, action)
 }
+
+// AddPrefix4ShardIndexes add expression prefix for shard index. e.g. an index is test.uk(tidb_shard(a), a).
+// It transforms the sql "SELECT * FROM test WHERE a = 10" to
+// "SELECT * FROM test WHERE tidb_shard(a) = val AND a = 10", val is the value of tidb_shard(10).
+// It also transforms the sql "SELECT * FROM test WHERE a IN (10, 20, 30)" to
+// "SELECT * FROM test WHERE tidb_shard(a) = val1 AND a = 10 OR tidb_shard(a) = val2 AND a = 20"
+// @param[in] conds            the original condtion of this datasource
+// @retval - the new condition after adding expression prefix
+func (ds *DataSource) AddPrefix4ShardIndexes(sc sessionctx.Context, conds []expression.Expression) []expression.Expression {
+	if !ds.containExprPrefixUk {
+		return conds
+	}
+
+	var err error
+	newConds := conds
```

Comment:
This assignment would not change the memory address.
If we want a new slice pointing to a new memory address, we need to `make([]Expression)` and `copy(src, dest)`.

Reply:
I do it to avoid unnecessary memory allocation and copy if it doesn't need add a prefix condition that was checked at the function call `ds.addExprPrefixCond`. If it need add a prefix condition, I make a new slice and copy the old to it. The address of `conds` was not changed.